internal/worker/sshsession/worker.go

```diff
@@ -0,0 +1,326 @@
+//     back from the controller's SSH server.
+//  5. Pipes the connection to the local sshd.
+//  6. On connection close, removes the ephemeral public key.
+func (w *sshSessionWorker) handleConnection(ctx context.Context, connID string) error {
+	reqParams, err := w.facadeClient.GetSSHConnRequest(connID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	ctrlAddress := reqParams.ControllerAddresses.Values()[0]
+	ephemeralPublicKey := string(reqParams.EphemeralPublicKey)
+
+	if err := w.ephemeralKeysUpdater.AddEphemeralKey(ephemeralPublicKey); err != nil {
+		return errors.Trace(err)
+	}
+
+	defer func() {
+		if err := w.ephemeralKeysUpdater.RemoveEphemeralKey(ephemeralPublicKey); err != nil {
+			w.logger.Errorf("Error cleaning up ephemeral public key: %v", err)
+		}
+	}()
+
+	if err := w.pipeConnectionToSSHD(ctx, ctrlAddress, reqParams.Password); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// pipeConnectionToSSHD initiates the connection back to the controller and pipes
+// it over to the local SSHD. This call blocks until the connection has finished.
+func (w *sshSessionWorker) pipeConnectionToSSHD(ctx context.Context, ctrlAddress, password string) error {
+	controllerConn, err := w.connectionGetter.GetControllerConnection(password, ctrlAddress)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	sshdConn, err := w.connectionGetter.GetSSHDConnection()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	// We make sure to close the connections when the context is done.
+	doneChan := make(chan struct{})
+	defer close(doneChan)
+	go func() {
+		select {
+		case <-ctx.Done():
+			controllerConn.Close()
+			sshdConn.Close()
+		case <-doneChan:
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		// sshd -> conn
+		defer controllerConn.Close()
+		defer sshdConn.Close()
+		_, _ = io.Copy(sshdConn, controllerConn)
+
+	}()
+
+	go func() {
+		defer wg.Done()
+		// conn -> sshd
+		defer controllerConn.Close()
+		defer sshdConn.Close()
+		_, _ = io.Copy(controllerConn, sshdConn)
+	}()
+	wg.Wait()
+	return nil
+}
+
+// ConnectionGetter provides the methods to connect to a controller and the local SSHD.
+type ConnectionGetter interface {
+	GetControllerConnection(password, ctrlAddress string) (net.Conn, error)
+	GetSSHDConnection() (net.Conn, error)
+}
+
+// connectionGetter is capable of initating SSH connections to two places.
+//
+//  1. The controller's SSH server
+//  2. The local SSHD on the machine
+//
+// The consumer is expect to pipe these connections together.
+type connectionGetter struct {
+	logger Logger
+}
+
+// newConnectionGetter returns a new connectionGetter.
+func newConnectionGetter(l Logger) *connectionGetter {
+	return &connectionGetter{l}
+}
+
+// GetControllerConnection initiates an SSH connection to the target ctrlAddress.
+func (w *connectionGetter) GetControllerConnection(password, ctrlAddress string) (net.Conn, error) {
```

Comment:
The `connectionGetter` struct and its methods don't seem to have any test coverage. Let's fix that in a follow-up. 
I also don't think this file needs to be split.